jib-core/src/main/java/com/google/cloud/tools/jib/builder/steps/PushLayersStep.java

```diff
@@ -76,34 +75,40 @@
   public ImmutableList<AsyncStep<PushBlobStep>> call() throws ExecutionException {
     try (TimerEventDispatcher ignored =
         new TimerEventDispatcher(buildConfiguration.getEventDispatcher(), DESCRIPTION)) {
-      ImmutableList<? extends AsyncStep<? extends CachedLayer>> cachedLayer =
+      ImmutableList<? extends AsyncStep<? extends CachedLayer>> cachedLayers =
           NonBlockingSteps.get(cachedLayerStep);
 
-      // Constructs a PushBlobStep for each layer.
-      ImmutableList.Builder<AsyncStep<PushBlobStep>> pushBlobStepsBuilder = ImmutableList.builder();
-      for (AsyncStep<? extends CachedLayer> cachedLayerStep : cachedLayer) {
-        ListenableFuture<PushBlobStep> pushBlobStepFuture =
-            Futures.whenAllSucceed(cachedLayerStep.getFuture())
-                .call(() -> makePushBlobStep(cachedLayerStep), listeningExecutorService);
-        pushBlobStepsBuilder.add(() -> pushBlobStepFuture);
+      try (ProgressEventDispatcher progressEventDispatcher =
+          progressEventDispatcherFactory.create("push layers", cachedLayers.size())) {
+        // Constructs a PushBlobStep for each layer.
+        ImmutableList.Builder<AsyncStep<PushBlobStep>> pushBlobStepsBuilder =
+            ImmutableList.builder();
+        for (AsyncStep<? extends CachedLayer> cachedLayerStep : cachedLayers) {
+          ProgressEventDispatcher.Factory childProgressEventDispatcherFactory =
+              progressEventDispatcher.newChildProducer();
+          ListenableFuture<PushBlobStep> pushBlobStepFuture =
+              Futures.whenAllSucceed(cachedLayerStep.getFuture())
+                  .call(
+                      () -> makePushBlobStep(cachedLayerStep, childProgressEventDispatcherFactory),
+                      listeningExecutorService);
+          pushBlobStepsBuilder.add(() -> pushBlobStepFuture);
+        }
+
+        return pushBlobStepsBuilder.build();
       }
-
-      return pushBlobStepsBuilder.build();
     }
   }
 
-  private PushBlobStep makePushBlobStep(AsyncStep<? extends CachedLayer> cachedLayerStep)
+  private PushBlobStep makePushBlobStep(
+      AsyncStep<? extends CachedLayer> cachedLayerStep,
+      ProgressEventDispatcher.Factory progressEventDispatcherFactory)
       throws ExecutionException {
     CachedLayer cachedLayer = NonBlockingSteps.get(cachedLayerStep);
 
-    Allocation progressAllocation =
```

Comment:
Double-checking: was it incorrect to add a new child previously?

Reply:
Yea, because a new child should have been created with the allocation units equal to the number of blobs to push, so that each of the `PushBlobStep` takes up a single unit of that new child rather than the parent.